lib/windows/launcher/actions/desktopShortcutActions.js

```diff
@@ -212,19 +226,28 @@ function createShortcutForMacOS(app) {
         };
         const wflowContent = Mustache.render(wflowContentSource, wflowContentData);
 
+        const icnsPath = path.join(tmpAppTemplatePath, '/Contents/Resources/icon.icns');
         fileUtil.copyFromAsar(appTemplatePath, tmpAppTemplatePath, 5)
         .then(() => fileUtil.chmodDir(tmpAppTemplatePath, mode))
-        .then(() => new Promise((resolve, reject) => {
-            // Convert icns file for the app
-            const icnsPath = path.join(tmpAppTemplatePath, '/Contents/Resources/icon.icns');
-            const input = fs.readFileSync(app.iconPath);
-            const output = png2icons.PNG2ICNS(input, png2icons.HERMITE, false, 0);
-            if (output) {
-                fs.writeFileSync(icnsPath, output);
-                resolve();
+        .then(() => {
+            // TODO: This implementation is duplicated due to compatitbility, remove it later.
+            if (app.shortcutIconPath) {
+                const suffix = app.shortcutIconPath.substring(4);
+                if (suffix === 'icns') {
+                    return fileUtil.copy(app.shortcutIconPath, icnsPath);
+                }
             }
-            reject(new Error('Error occured while converting icon.png to icon.icns.'));
-        }))
+            return new Promise((resolve, reject) => {
+                // Convert icns file for the app
+                const input = fs.readFileSync(app.iconPath);
+                const output = png2icons.PNG2ICNS(input, png2icons.HERMITE, false, 0);
+                if (output) {
+                    fs.writeFileSync(icnsPath, output);
+                    resolve();
+                }
+                reject(new Error('Error occured while converting icon.png to icon.icns.'));
+            });
+        })
```

Comment:
Promise should not be necessary here, as all operations inside the promise are synchronous. Could rewrite to:

```
const input = fs.readFileSync(app.iconPath);
const output = png2icons.PNG2ICNS(input, png2icons.HERMITE, false, 0);
if (!output) {
    throw new Error('Error occured while converting icon.png to icon.icns.');
}
fs.writeFileSync(icnsPath, output);
```

Reply:
Agree.